lib/tls_syntax/include/tls/tls_syntax.h

```diff
@@ -29,24 +35,192 @@ class ReadError : public std::invalid_argument
   using parent::parent;
 };
 
+///
+/// Struct-building tools
+///
+
+// Use this macro to define struct serialization with minimal boilerplate
+#define TLS_SERIALIZABLE(...)                                                  \
+  static const bool _tls_serializable = true;                                  \
+  auto _tls_fields_r() { return std::forward_as_tuple(__VA_ARGS__); }          \
+  auto _tls_fields_w() const { return std::forward_as_tuple(__VA_ARGS__); }
+
+// If your struct contains nontrivial members (e.g., vectors), use this to
+// define traits for them.
+#define TLS_TRAITS(...)                                                        \
+  static const bool _tls_has_traits = true;                                    \
+  using _tls_traits = std::tuple<__VA_ARGS__>;
+
+template<typename T>
+struct is_serializable
+{
+  template<typename U>
+  static std::true_type test(decltype(U::_tls_serializable));
+
+  template<typename U>
+  static std::false_type test(...);
+
+  static const bool value = decltype(test<T>(true))::value;
+};
+
+template<typename T>
+struct has_traits
+{
+  template<typename U>
+  static std::true_type test(decltype(U::_tls_has_traits));
+
+  template<typename U>
+  static std::false_type test(...);
+
+  static const bool value = decltype(test<T>(true))::value;
+};
+
+///
+/// size_of
+///
+
+template<typename T>
+struct is_serializable;
+
+template<typename T>
+inline typename std::enable_if<!is_serializable<T>::value && !std::is_enum<T>::value,
+                          size_t>::type
+size_of(const T& val);
+
+template<typename T>
+constexpr
+  typename std::enable_if<is_serializable<T>::value && !has_traits<T>::value,
+                          size_t>::type
+  size_of(const T& obj);
+
+template<typename T>
+constexpr
+  typename std::enable_if<is_serializable<T>::value && has_traits<T>::value,
+                          size_t>::type
+  size_of(const T& obj);
+
+// Primitive sizes
+template<>
+constexpr size_t size_of(const bool&) { return 1; }
+
+template<>
+constexpr size_t size_of(const uint8_t&) { return 1; }
+
+template<>
+constexpr size_t size_of(const uint16_t&) { return 2; }
+
+template<>
+constexpr size_t size_of(const uint32_t&) { return 4; }
+
+template<>
+constexpr size_t size_of(const uint64_t&) { return 8; }
+
+// Array size
+template<typename T, size_t N>
+constexpr size_t size_of(const std::array<T, N>& data)
+{
+  auto out = size_t(0);
+  for (const auto& item : data) {
+    out += size_of(item);
+  }
+  return out;
+}
+
+// Optional size
+template<typename T>
+constexpr size_t size_of(const std::optional<T>& opt)
+{
+  if (!opt) {
+    return 1;
+  }
+
+  return 1 + size_of(opt::get(opt));
```

Comment:
Using ternary operator `?:` could give better performance.
Or considering adding attributes [[likely]]/[[unlikely]] would be useful here.